x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -148,17 +150,39 @@ public void analyzedPlan(LogicalPlan parsed, ActionListener<LogicalPlan> listene
 
     private <T> void preAnalyze(LogicalPlan parsed, BiFunction<IndexResolution, EnrichResolution, T> action, ActionListener<T> listener) {
         PreAnalyzer.PreAnalysis preAnalysis = preAnalyzer.preAnalyze(parsed);
-        enrichPolicyResolver.resolvePolicy(preAnalysis.policyNames, listener.delegateFailureAndWrap((l, enrichResolution) -> {
+        var unresolvedPolicies = preAnalysis.unresolvedEnriches.stream()
+            .map(e -> new EnrichPolicyResolver.UnresolvedPolicy((String) e.policyName().fold(), e.mode()))
+            .collect(Collectors.toSet());
+        final Set<String> targetClusters = enrichPolicyResolver.groupIndicesPerCluster(
+            preAnalysis.indices.stream()
+                .flatMap(t -> Arrays.stream(Strings.commaDelimitedListToStringArray(t.id().index())))
+                .toArray(String[]::new)
+        ).keySet();
+        enrichPolicyResolver.resolvePolicies(targetClusters, unresolvedPolicies, listener.delegateFailureAndWrap((l, enrichResolution) -> {
             // first we need the match_fields names from enrich policies and THEN, with an updated list of fields, we call field_caps API
             var matchFields = enrichResolution.resolvedEnrichPolicies()
-                .stream()                 // triggered
-                .map(EnrichPolicy::getMatchField)
+                .stream()
+                .map(ResolvedEnrichPolicy::matchField)
                 .collect(Collectors.toSet());
-            preAnalyzeIndices(
-                parsed,
-                l.delegateFailureAndWrap((ll, indexResolution) -> ll.onResponse(action.apply(indexResolution, enrichResolution))),
-                matchFields
-            );
+            preAnalyzeIndices(parsed, l.delegateFailureAndWrap((ll, indexResolution) -> {
+                if (indexResolution.isValid()) {
+                    Set<String> newClusters = enrichPolicyResolver.groupIndicesPerCluster(
+                        indexResolution.get().concreteIndices().toArray(String[]::new)
+                    ).keySet();
+                    // If new clusters appear when resolving the main indices, we need to resolve the enrich policies again
+                    // or exclude main concrete indices. Since this is rare, it's simpler to resolve the enrich policies again.
+                    // TODO: add a test for this
```

Comment:
I will add a test for this in a follow-up as the size of this PR is pretty large already.